pkg/notification/client.go

```diff
@@ -0,0 +1,159 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	NotificationConfigMap   = "ods-notification"
+	UrlProperty             = "url"
+	MethodProperty          = "method"
+	ContentTypeProperty     = "contentType"
+	RequestTemplateProperty = "requestTemplate"
+	NotifyOnStatusProperty  = "notifyOnStatus"
+)
+
+type Client struct {
+	clientConfig     ClientConfig
+	httpClient       *http.Client
+	kubernetesClient kubernetes.ClientInterface
+}
+
+type ClientConfig struct {
+	Namespace string
+	Logger    logging.LeveledLoggerInterface
+}
+
+type PipelineRunResult struct {
+	PipelineRunURL string
+	OverallStatus  string
+	ODSContext     *pipelinectxt.ODSContext
+}
+
+type notificationConfig struct {
+	url            string
+	method         string
+	contentType    string
+	notifyOnStatus []string
+	template       *template.Template
+}
+
+func NewClient(config ClientConfig, kubernetesClient kubernetes.ClientInterface) (*Client, error) {
+	if config.Logger == nil {
+		config.Logger = &logging.LeveledLogger{Level: logging.LevelError}
+	}
+
+	return &Client{
+		clientConfig:     config,
+		httpClient:       &http.Client{},
+		kubernetesClient: kubernetesClient,
+	}, nil
+}
+
+func (c Client) readNotificationConfig(ctxt context.Context) (*notificationConfig, error) {
+	cm, err := c.kubernetesClient.GetConfigMap(ctxt, NotificationConfigMap, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to load %s ConfigMap: %v", NotificationConfigMap, err)
+	}
+
+	url, ok := cm.Data[UrlProperty]
+	if !ok {
+		return nil, fmt.Errorf("%s doesn't specify '%s' property", NotificationConfigMap, UrlProperty)
+	}
+
+	method, ok := cm.Data[MethodProperty]
+	if !ok {
+		return nil, fmt.Errorf("%s doesn't specify '%s' property", NotificationConfigMap, MethodProperty)
+	}
+
+	contentType, ok := cm.Data[ContentTypeProperty]
+	if !ok {
+		return nil, fmt.Errorf("%s doesn't specify '%s' property", NotificationConfigMap, ContentTypeProperty)
+	}
+
+	notifyOnStatus, ok := cm.Data[NotifyOnStatusProperty]
+	if !ok {
+		return nil, fmt.Errorf("%s doesn't specifiy '%s' property", NotificationConfigMap, NotifyOnStatusProperty)
+	}
+
+	decoder := json.NewDecoder(strings.NewReader(notifyOnStatus))
+	var notificationStatusValues []string
+	err = decoder.Decode(&notificationStatusValues)
+	if err != nil && err != io.EOF {
+		return nil, fmt.Errorf("decoding notification status properties failed: %w", err)
+	}
+
+	text, ok := cm.Data[RequestTemplateProperty]
+	if !ok {
+		return nil, fmt.Errorf("%s doesn't specify '%s' property", NotificationConfigMap, RequestTemplateProperty)
+	}
+
+	requestTemplate, err := template.New("requestTemplate").Parse(text)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse requestTemplate template")
+	}
+
+	return &notificationConfig{
+		url,
+		method,
+		contentType,
+		notificationStatusValues,
+		requestTemplate,
+	}, nil
+}
+
+func skipNotification(status string, allowedStatusValues []string) bool {
```

Comment:
I feel like it would be better to expose this, and have the caller check this instead of calling `CallWebhook`, and then magically it might not actually call it.

I see it more like this:
```go
if nc.NotifyForStatus(status) {
  nc.CallWebhook(...)
}
```

Writing this I realise that this requires to seed the notification client with the content of the ConfigMap ... hmm. That might be a good thing though - you could also change the notification client to receive a `notificationConfig` - and then the K8s client is only needed if that is not given, filling the config by making an API call.